pkg/server/plugin/nodeattestor/aws/iid.go

```diff
@@ -99,27 +131,50 @@ func (p *IIDAttestorPlugin) Attest(stream nodeattestor.Attest_PluginStream) erro
 		return caws.AttestationStepError("base64 decoding the IID signature", err)
 	}
 
-	p.mtx.Lock()
-	defer p.mtx.Unlock()
-
-	err = rsa.VerifyPKCS1v15(p.awsCaCertPublicKey, crypto.SHA256, docHash[:], sigBytes)
+	err = rsa.VerifyPKCS1v15(c.awsCaCertPublicKey, crypto.SHA256, docHash[:], sigBytes)
 	if err != nil {
 		return caws.AttestationStepError("verifying the cryptographic signature", err)
 	}
 
+	// query AWS for additional information to verify?
+	if !c.SkipEC2Calling {
+		err = p.ec2Attestation(c, doc)
+		if err != nil {
+			return fmt.Errorf("failed aws ec2 attestation: %v", err)
+		}
+	}
+
+	spiffeID, err := caws.MakeSpiffeID(c.trustDomain, c.pathTemplate, doc)
+	if err != nil {
+		return fmt.Errorf("failed to create spiffe ID: %v", err)
+	}
+
+	resp := &nodeattestor.AttestResponse{
+		Valid:        true,
+		BaseSPIFFEID: spiffeID.String(),
+	}
+
+	return stream.Send(resp)
+}
+
+// perform attestation backed by returns from AWS EC2 call(s)
+// meant to be called as part of Attest, and so uses the config from that call
+// for consistency rather than fetching a fresher (potentially altered) config.
+// returns nil on success
+func (p *IIDAttestorPlugin) ec2Attestation(c *IIDAttestorConfig, doc caws.InstanceIdentityDocument) error {
 	var awsSession *session.Session
 
-	if p.secretAccessKey != "" && p.accessKeyId != "" {
-		creds := credentials.NewStaticCredentials(p.accessKeyId, p.secretAccessKey, "")
+	if c.SecretAccessKey != "" && c.AccessKeyID != "" {
+		creds := credentials.NewStaticCredentials(c.AccessKeyID, c.SecretAccessKey, "")
 		awsSession = session.Must(session.NewSession(&aws.Config{Credentials: creds, Region: &doc.Region}))
```

Comment:
i know this is existing code, but `session.Must` shouldn't be used. Instead we should return the error. Also, we should consider creating this client during Configure() (it can be preserved as an unexported member of the config struct)

Reply:
is this not planned on eventually being long-running, where the access keys could be updated without process restart?

Reply:
that's a fair point when the credentials file is being used. i'm ok initializing the session here.